File: #0629.k-inverse-pairs-array.cpp
class Solution {
public:
    int kInversePairs(int n, int k) {

        long long m = 1e9+7;
        vector<vector<int> > dp(n + 2,vector<int>(k + 2,0));   
        for (int i=1;i<=n;i++) {
            for (int j = 0; j <= k; j++) {
                if (j==0) {
                    dp[i][j] = 1;
                } else{
                    int sum = (
                        dp[i - 1][j] + m - ((j - i) >= 0 ? 
                                          dp[i - 1][j - i] : 0)
                    ) % m;  
                    dp[i][j] = (dp[i][j - 1] + sum) % m;
                }
            }
        }

        return (
            dp[n][k] + m - ((k - 1) >= 0 ? dp[n][k - 1] : 0)
        ) % m;
    }
};

File: #0629.k-inverse-pairs-array.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number}
 */
var kInversePairs = function(n, k) {
    const mod = 1e9 + 7;
    let dp = Array(k + 1).fill(0);
    for (let i = 0; i <= n; ++i) {
        let temp = Array(k + 1).fill(0);
        temp[0] = 1;
        for (let j = 1; j <= k; ++j) {
            let val = (dp[j] + mod - ((j - i) >= 0 ? dp[j - i] : 0)) % mod;
            temp[j] = (temp[j - 1] + val) % mod;
        }
        dp = temp;
    }
    return ((dp[k] + mod - (k > 0 ? dp[k - 1] : 0)) % mod);
};
